### DISCUSSION TITLE
Lesson 9 -----   [FAIL. Reason: InvalidConsumer()] testCantEnterRaffleWhenCalculating()

### DISCUSSION THREAD

**Discussion by justAWanderKid** [Cyfrin/foundry-full-course-cu#1728] [Q&A] [answered]
Heres my Raffle.sol:

`

// Layout of Contract:
// version
// imports
// errors
// interfaces, libraries, contracts
// declarations
// State variables
// Events
// Modifiers
// Functions

// Layout of Functions:
// constructor
// receive function (if exists)
// fallback function (if exists)
// external
// public
// internal
// private
// internal & private view & pure functions
// external & public view & pure functions

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.25;

/*
 * @title Sample Raffle Contract
 * @author Patrick Colons
 * @notice this Contract is For Creating Sample Raffle
 * @dev implements Chainlink VRFv2
 */

import {VRFCoordinatorV2Interface} from "@chainlink/contracts/src/v0.8/vrf/interfaces/VRFCoordinatorV2Interface.sol";
import {VRFConsumerBaseV2} from "@chainlink/contracts/src/v0.8/vrf/VRFConsumerBaseV2.sol";

contract Raffle is VRFConsumerBaseV2 {

    //Errors
    error Raffle__NotEnoughEthSent();
    error Raffle__TransferFailed();
    error Raffle__RaffleNotOpen();
    error Raffle__UpKeepNotNeeded(uint ContractBalance,uint NumberOfPlayers,uint RaffleStateAsNumber);
    //Declarations

    enum RaffleState {
        OPEN,
        CALCULATING
    }

    //State Variables
    RaffleState private s_raffleState;

    uint16 private constant REQUEST_CONFIRMATIONS = 3;
    uint32 private constant NUMWORDS = 1;

    uint private immutable i_entranceFee;
    uint private immutable i_interval;
    uint private s_lastTimeStamp;
    VRFCoordinatorV2Interface private immutable i_vrfCoordinator;
    bytes32 private immutable i_gaslane;
    uint64 private immutable i_SubscriptionID;
    uint32 private immutable i_callbackGasLimit;

    address payable[] public s_players;
    address payable private s_recentWinner;

    //Events
    event EnteredRaffle(address indexed player);
    event PickedWinner(address indexed winner);

    //Functions
    constructor(uint _entranceFee, uint interval, address _vrfCoordinator, bytes32 _gaslane,uint64 _SubscriptionID,uint32 _callbackGasLimit) VRFConsumerBaseV2(_vrfCoordinator) {
        i_entranceFee = _entranceFee;
        i_interval = interval;
        s_lastTimeStamp = block.timestamp;
        i_vrfCoordinator = VRFCoordinatorV2Interface(_vrfCoordinator);
        i_gaslane = _gaslane;
        i_SubscriptionID = _SubscriptionID;
        i_callbackGasLimit = _callbackGasLimit;
        s_raffleState = RaffleState.OPEN;
    }

    function enterRaffle() external payable {
        if (msg.value < i_entranceFee) {
            revert Raffle__NotEnoughEthSent();
        }
        if (s_raffleState != RaffleState.OPEN) {
            revert Raffle__RaffleNotOpen();
        }
        s_players.push(payable(msg.sender));
        emit EnteredRaffle(msg.sender);
    }

    function checkUpKeep(bytes memory /*checkData*/) public view returns (bool upKeepNeeded, bytes memory /* performData */) {
        bool TimeHasPassed = (block.timestamp - s_lastTimeStamp) >= i_interval;
        bool isOpen = RaffleState.OPEN == s_raffleState;
        bool hasBalance = address(this).balance > 0;
        bool hasPlayers =  s_players.length > 0;
        upKeepNeeded = (TimeHasPassed && isOpen && hasBalance && hasPlayers);
        return (upKeepNeeded, "0x0");
    }

    function performUpKeep(bytes calldata /* performData */) external {
        (bool upkeepNeeded, ) =  checkUpKeep("");
        if (!upkeepNeeded) {
            revert Raffle__UpKeepNotNeeded(address(this).balance,s_players.length,uint(s_raffleState));
        }
        s_raffleState = RaffleState.CALCULATING;
        /*uint requestID =*/ i_vrfCoordinator.requestRandomWords(i_gaslane,i_SubscriptionID,REQUEST_CONFIRMATIONS,i_callbackGasLimit,NUMWORDS);
    }

    function fulfillRandomWords(uint256 /*_requestID*/, uint256[] memory _randomWords) internal override {
        uint indexOfWinner = _randomWords[0] % s_players.length;
        address payable winner = s_players[indexOfWinner];
        s_recentWinner = winner;
        s_raffleState = RaffleState.OPEN;
        
        s_players = new address payable[](0);
        s_lastTimeStamp = block.timestamp;

        (bool success,) = winner.call{value: address(this).balance}("");
        if (!success) {
            revert Raffle__TransferFailed();
        }
        emit PickedWinner(winner);
    }

    //Getter Functions
    
    function getEntranceFee() external view returns(uint) {
        return i_entranceFee;
    }

    function getInterval() external view returns(uint) {
        return i_interval;
    }

    function getArrayLength() external view returns (uint) {
        return s_players.length;
    }

    
    function getPlayerAddress(uint _index) external view returns (address) {
        return s_players[_index];
    }

    function getRaffleState() external view returns (RaffleState) {
        return s_raffleState;
    }

}
`

my HelperConfig.s.sol:

`

// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity ^0.8.25;

import {Script} from "forge-std/Script.sol";
import {VRFCoordinatorV2Mock} from "@chainlink/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol";
import {LinkToken} from "../test/mocks/LinkToken.sol";

contract HelperConfig is Script {

    struct NetworkConfig {
        uint _entranceFee;
        uint interval;
        address _vrfCoordinator;
        bytes32 _gaslane;
        uint64 _SubscriptionID;
        uint32 _callbackGasLimit;
        address _LinkTokenAddr;
    }

    NetworkConfig public activeNetworkConfig;

    constructor() {
        if (block.chainid == 11155111) {
            activeNetworkConfig = getSepoliaEthConfig();
        } else if (block.chainid == 31337) {
            activeNetworkConfig = getOrCreateAnvilEthConfig();
        }
    }

    function getSepoliaEthConfig() public pure returns (NetworkConfig memory) {
        return NetworkConfig({
            _entranceFee:0.01 ether,
            interval:30,
            _vrfCoordinator:0x8103B0A8A00be2DDC778e6e7eaa21791Cd364625,
            _gaslane:0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c,
            _SubscriptionID:0,
            _callbackGasLimit:500000,
            _LinkTokenAddr:0x779877A7B0D9E8603169DdbD7836e478b4624789
        });
    }

    function getOrCreateAnvilEthConfig() public returns (NetworkConfig memory) {
        if (activeNetworkConfig._vrfCoordinator != address(0)) { 
            return activeNetworkConfig;
        }

        uint96 baseFee = 0.25 ether;
        uint96 gasPriceLink = 1e9;

        vm.startBroadcast();
        VRFCoordinatorV2Mock vrfCoordinatorV2Mock = new VRFCoordinatorV2Mock(baseFee,gasPriceLink);
        LinkToken linkToken = new LinkToken();
        vm.stopBroadcast();

        return NetworkConfig({
            _entranceFee:0.01 ether,
            interval:30,
            _vrfCoordinator:address(vrfCoordinatorV2Mock),
            _gaslane:0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c,
            _SubscriptionID:0,
            _callbackGasLimit:500000,
            _LinkTokenAddr: address(linkToken)
        });

    }

}
`

my DeployRaffle.s.sol:

`

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.25;

import {Script} from "forge-std/Script.sol";
import {Raffle} from "../src/Raffle.sol";
import {HelperConfig} from "./HelperConfig.s.sol";
import {CreateSubscription,FundSubscription,AddConsumer} from "./interactions.s.sol";

contract DeployRaffle is Script {

    function run() external returns (Raffle,HelperConfig) {

        HelperConfig helperConfig = new HelperConfig();
        (uint _entranceFee, uint interval, address _vrfCoordinator, bytes32 _gaslane, uint64 _SubscriptionID, uint32 _callbackGasLimit,address _LinkTokenAddr) = helperConfig.activeNetworkConfig();

        vm.startBroadcast();
        Raffle raffle = new Raffle(
            _entranceFee,
            interval,
            _vrfCoordinator,
            _gaslane,
            _SubscriptionID,
            _callbackGasLimit
            );
        vm.stopBroadcast();

        
        if (_SubscriptionID == 0) {
            CreateSubscription createSubscription = new CreateSubscription();
            _SubscriptionID = createSubscription.createSubscription(_vrfCoordinator);

            FundSubscription fundSubscription = new FundSubscription();
            fundSubscription.fundSubscription(_vrfCoordinator,_SubscriptionID,_LinkTokenAddr);

        }

        

        AddConsumer addConsumer = new AddConsumer();
        addConsumer.addConsumer(address(raffle),_vrfCoordinator,_SubscriptionID);

        
        return (raffle,helperConfig);
    }

}
`

my interactions.s.sol:

`

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.25;

import {Script,console} from "forge-std/Script.sol";
import {HelperConfig} from "./HelperConfig.s.sol";
import {VRFCoordinatorV2Mock} from "@chainlink/contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol";
import {LinkToken} from "../test/mocks/LinkToken.sol";
import {DevOpsTools} from "lib/foundry-devops/src/DevOpsTools.sol";

contract CreateSubscription is Script {

    function CreateSubscriptionUsingConfig() public returns (uint64) {
        HelperConfig helperConfig = new HelperConfig();
        (,, address _vrfCoordinator,,,,) = helperConfig.activeNetworkConfig();
        return createSubscription(_vrfCoordinator);
    }

    function createSubscription(address ___vrfCoordinator) public returns (uint64) {
        console.log("Creating Subscription on Chain ID: ", block.chainid);
        vm.startBroadcast();
        uint64 subID = VRFCoordinatorV2Mock(___vrfCoordinator).createSubscription();
        vm.stopBroadcast();
        console.log("Your SubID is: ", subID);
        console.log("Please Update Subscription ID in HelperConfig.s.sol");
        return subID;
    }

    function run() external returns (uint64) {
        return CreateSubscriptionUsingConfig();
    }

}

contract FundSubscription is Script {

    uint96 public constant FUND_AMOUNT = 3 ether;

    function fundSubscriptionUsingConfig() public {
        HelperConfig helperConfig = new HelperConfig();
        (,, address _vrfCoordinator,,uint64 _SubscriptionID,,address _LinkTokenAddr) = helperConfig.activeNetworkConfig();
        fundSubscription(_vrfCoordinator,_SubscriptionID,_LinkTokenAddr);
    }

    function fundSubscription(address ___vrfCoordinator,uint64 ___SubscriptionID,address ___LinkTokenAddr) public {
        console.log("Funding Subscription ", ___SubscriptionID);
        console.log("Using vrfCoordinator ", ___vrfCoordinator);
        console.log("On Chain ID: ", block.chainid);
        if (block.chainid == 31337) {
            vm.startBroadcast();
            VRFCoordinatorV2Mock(___vrfCoordinator).fundSubscription(___SubscriptionID,FUND_AMOUNT);
            vm.stopBroadcast();
        } else {
            vm.startBroadcast();
            LinkToken(___LinkTokenAddr).transferAndCall(___vrfCoordinator, FUND_AMOUNT,abi.encode(___SubscriptionID));
            vm.startBroadcast();
        }
    }

    function run() external {
        fundSubscriptionUsingConfig();
    }

}

contract AddConsumer is Script {

    function addConsumer(address ___raffle, address ___vrfCoordinator, uint64 ___SubscriptionID) public {
        console.log("Adding Consumer Contract: ", ___raffle);
        console.log("Using vrfCoordinator: ", ___vrfCoordinator);
        console.log("On ChainID: ", block.chainid);
        vm.startBroadcast();
        VRFCoordinatorV2Mock(___vrfCoordinator).addConsumer(___SubscriptionID,___raffle);
        vm.stopBroadcast();

    }

    function addConsumerUsingConfig(address _raffle) public {
        HelperConfig helperConfig = new HelperConfig();
        (,, address _vrfCoordinator,,uint64 _SubscriptionID,,) = helperConfig.activeNetworkConfig();
        addConsumer(_raffle,_vrfCoordinator,_SubscriptionID);

    }

    function run() external {
        address raffle = DevOpsTools.get_most_recent_deployment("Raffle", block.chainid);
        addConsumerUsingConfig(raffle);
    }
}

`

and finally RaffleTest.t.sol:

`

// SPDX-License-Identifier: MIT
pragma solidity ^0.8.25;

import {Test,console} from "forge-std/Test.sol";
import {DeployRaffle} from "../../script/DeployRaffle.s.sol";
import {Raffle} from "../../src/Raffle.sol";
import {HelperConfig} from "../../script/HelperConfig.s.sol";

contract RaffleTest is Test {

    event EnteredRaffle(address indexed player);

    Raffle raffle;
    HelperConfig helperConfig;

    uint _entranceFee;
    uint interval;
    address _vrfCoordinator;
    bytes32 _gaslane;
    uint64 _SubscriptionID;
    uint32 _callbackGasLimit;
    address _LinkTokenAddr;

    address public Player = makeAddr("Player");
    uint constant public STARTING_BALANCE = 10 ether;

    

    function setUp() external {
        DeployRaffle deployer = new DeployRaffle();
        (raffle,helperConfig) = deployer.run();
        (
            _entranceFee,
            interval,
            _vrfCoordinator,
            _gaslane,
            _SubscriptionID,
            _callbackGasLimit,
            _LinkTokenAddr
        ) = helperConfig.activeNetworkConfig();
        vm.deal(Player,STARTING_BALANCE);
    }

    function testRaffleInitializesInOpenState() public view {
        assert(raffle.getRaffleState() == Raffle.RaffleState.OPEN);
    }

    function testRaffleRevertsWhenYouDontPayEnough() public  {
        //arrange
        vm.prank(Player);
        //Account
        vm.expectRevert(Raffle.Raffle__NotEnoughEthSent.selector);
        //assert
        raffle.enterRaffle();
    }

    function testRaffleRecordsPlayerWhenTheyEnter() public {
        vm.prank(Player);
        raffle.enterRaffle{value:_entranceFee}();
        address PlayerAddr = raffle.getPlayerAddress(0);
        assertEq(PlayerAddr,Player);
    }

    function testEmitsEventOnEntrance() public {
        vm.prank(Player);
        vm.expectEmit(true,false,false,false, address(raffle));
        emit Raffle.EnteredRaffle(Player);
        raffle.enterRaffle{value:_entranceFee}();
    }

    function testCantEnterRaffleWhenCalculating() public {
        
        vm.prank(Player);
        raffle.enterRaffle{value:_entranceFee}();

        vm.warp(block.timestamp + interval + 1);
        vm.roll(block.number + 1);

        raffle.performUpKeep("");

        vm.expectRevert(Raffle.Raffle__RaffleNotOpen.selector);
        vm.prank(Player);

        raffle.enterRaffle{value:_entranceFee}();
    }

}
`

if i run `forge test --match-test testCantEnterRaffleWhenCalculating -vvvvvv` in terminal, this is the results i get:

`
[⠊] Compiling...
[⠃] Compiling 41 files with Solc 0.8.25
[⠊] Solc 0.8.25 finished in 3.75s
Compiler run successful!

Ran 1 test for test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: InvalidConsumer()] testCantEnterRaffleWhenCalculating() (gas: 101937)
Logs:
  Creating Subscription on Chain ID:  31337
  Your SubID is:  1
  Please Update Subscription ID in HelperConfig.s.sol
  Funding Subscription  1
  Using vrfCoordinator  0x90193C961A926261B756D1E5bb255e67ff9498A1
  On Chain ID:  31337
  Adding Consumer Contract:  0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
  Using vrfCoordinator:  0x90193C961A926261B756D1E5bb255e67ff9498A1
  On ChainID:  31337

Traces:
  [46542270] RaffleTest::setUp()
    ├─ [23293270] → new DeployRaffle@0x5615dEB798BB3E4dFa0139dFa1b3D433Cc23b72f
    │   └─ ← [Return] 116118 bytes of code
    ├─ [23026436] DeployRaffle::run()
    │   ├─ [4924801] → new HelperConfig@0x104fBc016F4bb334D775a19E8A6510109AC63E00
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [1393898] → new VRFCoordinatorV2Mock@0x90193C961A926261B756D1E5bb255e67ff9498A1
    │   │   │   ├─ emit ConfigSet()
    │   │   │   └─ ← [Return] 6612 bytes of code
    │   │   ├─ [746641] → new LinkToken@0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496
    │   │   │   ├─ emit Transfer(from: 0x0000000000000000000000000000000000000000, to: DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38], amount: 1000000000000000000000000 [1e24])
    │   │   │   └─ ← [Return] 3269 bytes of code
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Return] 12897 bytes of code
    │   ├─ [963] HelperConfig::activeNetworkConfig() [staticcall]
    │   │   └─ ← [Return] 10000000000000000 [1e16], 30, VRFCoordinatorV2Mock: [0x90193C961A926261B756D1E5bb255e67ff9498A1], 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, 0, 500000 [5e5], LinkToken: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496]
    │   ├─ [0] VM::startBroadcast()
    │   │   └─ ← [Return] 
    │   ├─ [600229] → new Raffle@0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3
    │   │   └─ ← [Return] 2873 bytes of code
    │   ├─ [0] VM::stopBroadcast()
    │   │   └─ ← [Return] 
    │   ├─ [5454867] → new CreateSubscription@0x037eDa3aDB1198021A9b2e88C22B464fD38db3f3
    │   │   └─ ← [Return] 27131 bytes of code
    │   ├─ [53085] CreateSubscription::createSubscription(VRFCoordinatorV2Mock: [0x90193C961A926261B756D1E5bb255e67ff9498A1])
    │   │   ├─ [0] console::log("Creating Subscription on Chain ID: ", 31337 [3.133e4]) [staticcall]
    │   │   │   └─ ← [Stop] 
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [46829] VRFCoordinatorV2Mock::createSubscription()
    │   │   │   ├─ emit SubscriptionCreated(subId: 1, owner: DefaultSender: [0x1804c8AB1F12E6bbf3894d4083f33e07309d1f38])
    │   │   │   └─ ← [Return] 1
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [0] console::log("Your SubID is: ", 1) [staticcall]
    │   │   │   └─ ← [Stop] 
    │   │   ├─ [0] console::log("Please Update Subscription ID in HelperConfig.s.sol") [staticcall]
    │   │   │   └─ ← [Stop] 
    │   │   └─ ← [Return] 1
    │   ├─ [5549407] → new FundSubscription@0xDDc10602782af652bB913f7bdE1fD82981Db7dd9
    │   │   └─ ← [Return] 27603 bytes of code
    │   ├─ [6970] FundSubscription::fundSubscription(VRFCoordinatorV2Mock: [0x90193C961A926261B756D1E5bb255e67ff9498A1], 1, LinkToken: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496])
    │   │   ├─ [0] console::log("Funding Subscription ", 1) [staticcall]
    │   │   │   └─ ← [Stop] 
    │   │   ├─ [0] console::log("Using vrfCoordinator ", VRFCoordinatorV2Mock: [0x90193C961A926261B756D1E5bb255e67ff9498A1]) [staticcall]
    │   │   │   └─ ← [Stop] 
    │   │   ├─ [0] console::log("On Chain ID: ", 31337 [3.133e4]) [staticcall]
    │   │   │   └─ ← [Stop] 
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [3178] VRFCoordinatorV2Mock::fundSubscription(1, 3000000000000000000 [3e18])
    │   │   │   ├─ emit SubscriptionFunded(subId: 1, oldBalance: 0, newBalance: 3000000000000000000 [3e18])
    │   │   │   └─ ← [Stop] 
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Stop] 
    │   ├─ [6205783] → new AddConsumer@0x7FdB3132Ff7D02d8B9e221c61cC895ce9a4bb773
    │   │   └─ ← [Return] 30880 bytes of code
    │   ├─ [51034] AddConsumer::addConsumer(Raffle: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3], VRFCoordinatorV2Mock: [0x90193C961A926261B756D1E5bb255e67ff9498A1], 1)
    │   │   ├─ [0] console::log("Adding Consumer Contract: ", Raffle: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3]) [staticcall]
    │   │   │   └─ ← [Stop] 
    │   │   ├─ [0] console::log("Using vrfCoordinator: ", VRFCoordinatorV2Mock: [0x90193C961A926261B756D1E5bb255e67ff9498A1]) [staticcall]
    │   │   │   └─ ← [Stop] 
    │   │   ├─ [0] console::log("On ChainID: ", 31337 [3.133e4]) [staticcall]
    │   │   │   └─ ← [Stop] 
    │   │   ├─ [0] VM::startBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   ├─ [47240] VRFCoordinatorV2Mock::addConsumer(1, Raffle: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3])
    │   │   │   ├─ emit ConsumerAdded(subId: 1, consumer: Raffle: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3])
    │   │   │   └─ ← [Stop] 
    │   │   ├─ [0] VM::stopBroadcast()
    │   │   │   └─ ← [Return] 
    │   │   └─ ← [Stop] 
    │   └─ ← [Return] Raffle: [0xBb2180ebd78ce97360503434eD37fcf4a1Df61c3], HelperConfig: [0x104fBc016F4bb334D775a19E8A6510109AC63E00]
    ├─ [963] HelperConfig::activeNetworkConfig() [staticcall]
    │   └─ ← [Return] 10000000000000000 [1e16], 30, VRFCoordinatorV2Mock: [0x90193C961A926261B756D1E5bb255e67ff9498A1], 0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, 0, 500000 [5e5], LinkToken: [0xA8452Ec99ce0C64f20701dB7dD3abDb607c00496]
    ├─ [0] VM::deal(Player: [0xF70E4B142b9568324c3Ea8f28cb073B9bd1e1d3a], 10000000000000000000 [1e19])
    │   └─ ← [Return] 
    └─ ← [Stop] 

  [101937] RaffleTest::testCantEnterRaffleWhenCalculating()
    ├─ [0] VM::prank(Player: [0xF70E4B142b9568324c3Ea8f28cb073B9bd1e1d3a])
    │   └─ ← [Return] 
    ├─ [47809] Raffle::enterRaffle{value: 10000000000000000}()
    │   ├─ emit EnteredRaffle(player: Player: [0xF70E4B142b9568324c3Ea8f28cb073B9bd1e1d3a])
    │   └─ ← [Stop] 
    ├─ [0] VM::warp(32)
    │   └─ ← [Return] 
    ├─ [0] VM::roll(2)
    │   └─ ← [Return] 
    ├─ [31425] Raffle::performUpKeep(0x)
    │   ├─ [5271] VRFCoordinatorV2Mock::requestRandomWords(0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c, 0, 3, 500000 [5e5], 1)
    │   │   └─ ← [Revert] InvalidConsumer()
    │   └─ ← [Revert] InvalidConsumer()
    └─ ← [Revert] InvalidConsumer()

Suite result: FAILED. 0 passed; 1 failed; 0 skipped; finished in 3.94ms (91.00µs CPU time)

Ran 1 test suite in 7.05s (3.94ms CPU time): 0 tests passed, 1 failed, 0 skipped (1 total tests)

Failing tests:
Encountered 1 failing test in test/unit/RaffleTest.t.sol:RaffleTest
[FAIL. Reason: InvalidConsumer()] testCantEnterRaffleWhenCalculating() (gas: 101937)

Encountered a total of 1 failing tests, 0 tests succeeded
`

i took a Look at the Function that reverts the InvalidConsumer() error, i couldn't understand what's the Problem Causing this. 

i Appreciate all the Helps im Gonna get :)

**Comment by EngrPips** [comment]
Hello @tornado06, below is my `DeployLottery` script

```js
// SPDX-License-Identifier: UNLICENSED
pragma solidity 0.8.23;
import {Script, console2} from "forge-std/Script.sol";
import {HelperConfig} from "./HelperConfig.s.sol";
import {Lottery} from "../src/Lottery.sol";
import {CreateSubscription} from "../script/Interactions/CreateSubscription.s.sol";
import {FundSubscription} from "../script/Interactions/FundSubscription.s.sol";
import {AddConsumer} from "../script/Interactions/AddConsumer.s.sol";

contract DeployLottery is Script {

    uint256 _entryFee;
    uint256 _interval;
    address _vrfCoordinatorContractAddress;
    bytes32 _keyHash;
    uint64 _subscriptionId;
    address _linkToken;
    uint256 _deployerKey;
    
    function run() external returns(Lottery, HelperConfig){
        HelperConfig helperConfig = new HelperConfig();
        (_entryFee, _interval, _vrfCoordinatorContractAddress, _keyHash, _subscriptionId, _linkToken, _deployerKey) = helperConfig.activeNetworkConfig();
        console2.log("The subscriptionIs is: ", _subscriptionId);

        if(_subscriptionId == 0) {
            CreateSubscription createSubscriptionScript = new CreateSubscription();

            
            _subscriptionId = createSubscriptionScript.createSubscription(_vrfCoordinatorContractAddress, _deployerKey);
            

            FundSubscription fundSubscriptionScript = new FundSubscription();

            
            fundSubscriptionScript.fundSubscription(_subscriptionId, _linkToken, _vrfCoordinatorContractAddress, _deployerKey);
            
        }

        vm.startBroadcast();
        Lottery lotteryContract = new Lottery(_entryFee, _interval, _vrfCoordinatorContractAddress, _keyHash, _subscriptionId);
        vm.stopBroadcast();

        AddConsumer addConsumerScript = new AddConsumer();

        
        addConsumerScript.addConsumer(_subscriptionId,address(lotteryContract), _vrfCoordinatorContractAddress, _deployerKey);
        

        return (lotteryContract, helperConfig);
    }
}
```

I notice one difference between my `DeployLottery` script and yours which is you are creating the Lottery contract before creating a subscription on chainlink VRF so my question how do you set the `i_subscriptionId;` on your Lottery contract.

**Reply by EngrPips** [reply]
What makes sense most is you are setting the `i_subscriptionId` in your Lottery to the `0` which you get from your `HelperConfig` script 

```js
function getOrCreateAnvilEthConfig() public returns (NetworkConfig memory) {
    if (activeNetworkConfig._vrfCoordinator != address(0)) { 
        return activeNetworkConfig;
    }

    uint96 baseFee = 0.25 ether;
    uint96 gasPriceLink = 1e9;

    vm.startBroadcast();
    VRFCoordinatorV2Mock vrfCoordinatorV2Mock = new VRFCoordinatorV2Mock(baseFee,gasPriceLink);
    LinkToken linkToken = new LinkToken();
    vm.stopBroadcast();

    return NetworkConfig({
        _entranceFee:0.01 ether,
        interval:30,
        _vrfCoordinator:address(vrfCoordinatorV2Mock),
        _gaslane:0x474e34a077df58807dbe9c96d3c009b23b3c6d0cce433e59bbf5b34f823bc56c,
        _SubscriptionID:0, <= here
        _callbackGasLimit:500000,
        _LinkTokenAddr: address(linkToken)
    });
```

**Reply by justAWanderKid** [reply, marked as answer]
God Damn Man, i Hate it When i Make Such Silly Mistakes!

you're right, i need to create Subscription ID before Deploying the Raffle Contract!


my old DeployRaffle.s.sol Which Caused an Error When Running test:

`


// SPDX-License-Identifier: MIT
pragma solidity ^0.8.25;


import {Script} from "forge-std/Script.sol";
import {Raffle} from "../src/Raffle.sol";
import {HelperConfig} from "./HelperConfig.s.sol";
import {CreateSubscription,FundSubscription,AddConsumer} from "./interactions.s.sol";

contract DeployRaffle is Script {

    function run() external returns (Raffle,HelperConfig) {

        HelperConfig helperConfig = new HelperConfig();
        (uint _entranceFee, uint interval, address _vrfCoordinator, bytes32 _gaslane, uint64 _SubscriptionID, uint32 _callbackGasLimit,address _LinkTokenAddr) = helperConfig.activeNetworkConfig();

        vm.startBroadcast();
        Raffle raffle = new Raffle(
            _entranceFee,
            interval,
            _vrfCoordinator,
            _gaslane,
            _SubscriptionID,
            _callbackGasLimit
            );
        vm.stopBroadcast();

        
        if (_SubscriptionID == 0) {
            CreateSubscription createSubscription = new CreateSubscription();
            _SubscriptionID = createSubscription.createSubscription(_vrfCoordinator);

            FundSubscription fundSubscription = new FundSubscription();
            fundSubscription.fundSubscription(_vrfCoordinator,_SubscriptionID,_LinkTokenAddr);

        }

        

        AddConsumer addConsumer = new AddConsumer();
        addConsumer.addConsumer(address(raffle),_vrfCoordinator,_SubscriptionID);

        
        return (raffle,helperConfig);
    }

}
`

and the Change i Made to it, Which successfully Ran the Test Without an Error:

`


// SPDX-License-Identifier: MIT
pragma solidity ^0.8.25;


import {Script} from "forge-std/Script.sol";
import {Raffle} from "../src/Raffle.sol";
import {HelperConfig} from "./HelperConfig.s.sol";
import {CreateSubscription,FundSubscription,AddConsumer} from "./interactions.s.sol";

contract DeployRaffle is Script {

    function run() external returns (Raffle,HelperConfig) {

        HelperConfig helperConfig = new HelperConfig();
        (uint _entranceFee, uint interval, address _vrfCoordinator, bytes32 _gaslane, uint64 _SubscriptionID, uint32 _callbackGasLimit,address _LinkTokenAddr) = helperConfig.activeNetworkConfig();

        if (_SubscriptionID == 0) {
            CreateSubscription createSubscription = new CreateSubscription();
            _SubscriptionID = createSubscription.createSubscription(_vrfCoordinator);

            FundSubscription fundSubscription = new FundSubscription();
            fundSubscription.fundSubscription(_vrfCoordinator,_SubscriptionID,_LinkTokenAddr);

        }
        

        vm.startBroadcast();
        Raffle raffle = new Raffle(
            _entranceFee,
            interval,
            _vrfCoordinator,
            _gaslane,
            _SubscriptionID,
            _callbackGasLimit
            );
        vm.stopBroadcast();

        AddConsumer addConsumer = new AddConsumer();
        addConsumer.addConsumer(address(raffle),_vrfCoordinator,_SubscriptionID);

        
        return (raffle,helperConfig);
    }

}
`



i Appreciate Your help mate!

**Reply by EngrPips** [reply]
You are welcome
